File: C/Algorithm/Algorithm I/3. Longest Substring Without Repeating Characters.c
/*
 * Calculates the length of the longest substring without repeating characters.
 *
 * The 'lengthOfLongestSubstring' function calculates the length of the longest substring
 * in a given string 's' that does not contain any repeating characters. It iterates
 * through the string using a sliding window approach, maintaining a substring defined by
 * 'substringStartIndex' and 'substringEndIndex'. It uses an integer array 'currentSubstring'
 * to keep track of the frequency of characters in the current substring. If a character is
 * already present in the current substring, the function slides the window by incrementing
 * 'substringStartIndex' until the repeating character is removed from the substring. The
 * function also updates the 'maxSubstringLength' variable with the length of the longest
 * non-repeating substring encountered during the iteration.
 *
 * Parameters:
 * - s: A pointer to a null-terminated string.
 *
 * Returns:
 * The length of the longest substring without repeating characters.
 *
 * This function efficiently calculates the length of the longest non-repeating substring by
 * utilizing a sliding window approach and keeping track of character frequencies. It returns
 * the length of the longest substring found in the input string 's'.
 */
int lengthOfLongestSubstring(char * s) {
    int stringLength = strlen(s);
    int maxSubstringLength = 0;
    int substringStartIndex = 0;

    int currentSubstring[512];
    memset(currentSubstring, 0, sizeof(int) * 512);
    
    for (int substringEndIndex = 0; substringEndIndex < stringLength; substringEndIndex++) {
        int charIndex = s[substringEndIndex];

        if (currentSubstring[charIndex]) {
            assert(currentSubstring[charIndex] == 1);

            while (substringStartIndex < substringEndIndex) {
                int decreaseIndex = s[substringStartIndex];
                currentSubstring[decreaseIndex]--;
                substringStartIndex++;

                if (currentSubstring[charIndex] == 0)
                    break;
            }
        }

        currentSubstring[charIndex]++;
        maxSubstringLength = MAX(maxSubstringLength, substringEndIndex - substringStartIndex + 1);
    }

    return maxSubstringLength;
}

File: Python/Algorithm/Algorithm I/3. Longest Substring Without Repeating Characters.py
import collections

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        Find the length of the longest substring without repeating characters.

        Args:
            s (str): Input string.

        Returns:
            int: Length of the longest substring without repeating characters.

        Time Complexity: o(n^2) where n is the length of the input string.
        Space Complexity: o(n) where n is the length of the input string.
        """
        longestSubstring = 0
                
        currentWindow = collections.OrderedDict()    
        
        for index, char in enumerate(s):
            while char in currentWindow:
                currentWindow.popitem(last=False)
            
            longestSubstring = max(longestSubstring, len(currentWindow) + 1)
            currentWindow[char] = index

        return longestSubstring
